include/envoy/stats/refcount_ptr.h

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include <atomic>
+
+#include "envoy/common/pure.h"
+
+namespace Envoy {
+namespace Stats {
+
+// Implements a reference-counted pointer to a class, so that its external usage
```

Comment:
I think you already asked this of @ahedberg but it doesn't thrill me that we are implementing this. Surely there is somewhere we can get this from that is already used in production?

Reply:
There is existing `gpr_refcount` struct in gRPC C core doing similar things though I'm not very sure it will help the memory usage here.

Reply:
Thanks for the super detailed explanation. Let me have some time to digest all of this today/tomorrow and I will get back to you, but I don't have any objection to shipping this if we think it's the right thing to do. It's conceptually not that hard to understand I agree.

Reply:
sg; no rush; if there's anything else I can do to make it clearer and easy to reason about happy to do it.

Reply:
> Do you have any more detail on whether/why that would be a good way to go?

I don't have any real reason for using the grpc::RefCount, I didn't know it had a C++ version either (was looking at C version gpr_refcount). It is just one of a implementation that is already used in production that Matt asked, and it is already one of our dependencies. Though I think it does pretty similar things to here?

Reply:
The only thing I'll add to this discussion is that http://wg21.link/p0406 had overall positive feedback at the standards committee meeting in Kona in 2017, but hasn't been seen again as far as I can tell. That won't help us until post-C++20 at this point, but it is a signal that the standards committee has a) thought about this and b) doesn't have anything to do what we want to do here yet.

Reply:
In parallel is it worth it to open an issue with absl and request they support something like this with a link to our code? Perhaps they would be willing to take over ownership at some point and we could then depend on absl for this?

Reply:
Worth discussing with them, yes. There's a similar class internally that we may be able to convince Abseil to release.

Reply:
@jmarantz can you potentially open an issue there and get their guidance?

Reply:
https://github.com/abseil/abseil-cpp/issues/344